src/Microsoft.NetCore.Analyzers/Core/Performance/UsePropertyInsteadOfCountMethodWhenAvailable.cs

```diff
@@ -0,0 +1,226 @@
+            helpLinkUri: "https://docs.microsoft.com/visualstudio/code-quality/" + RuleId.ToLowerInvariant());
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+        /// <summary>
+        /// Returns a set of descriptors for the diagnostics that this analyzer is capable of producing.
+        /// </summary>
+        /// <value>The supported diagnostics.</value>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(s_rule);
+
+        /// <summary>
+        /// Called once at session start to register actions in the analysis context.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        /// <summary>
+        /// Called on compilation start.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (WellKnownTypes.Enumerable(context.Compilation) is INamedTypeSymbol enumerableType)
+            {
+                var operationActionsContext = new UsePropertyInsteadOfCountMethodWhenAvailableAnalyzer.OperationActionsContext(
+                    context.Compilation,
+                    enumerableType);
+
+                context.RegisterOperationAction(
+                    CreateOperationActionsHandler(operationActionsContext).AnalyzeInvocationOperation,
+                    OperationKind.Invocation);
+            }
+        }
+
+        /// <summary>
+        /// Creates the operation actions handler.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <returns>The operation actions handler.</returns>
+        protected abstract OperationActionsHandler CreateOperationActionsHandler(OperationActionsContext context);
+
+        protected sealed class OperationActionsContext
+        {
+            private readonly Lazy<INamedTypeSymbol> _immutableArrayType;
+            private readonly Lazy<IPropertySymbol> _iCollectionCountProperty;
+            private readonly Lazy<INamedTypeSymbol> _iCollectionOfType;
+
+            public OperationActionsContext(Compilation compilation, INamedTypeSymbol enumerableType)
+            {
+                Compilation = compilation;
+                EnumerableType = enumerableType;
+                _immutableArrayType = new Lazy<INamedTypeSymbol>(() => WellKnownTypes.ImmutableArray(Compilation), true);
+                _iCollectionCountProperty = new Lazy<IPropertySymbol>(() => WellKnownTypes.ICollection(Compilation)?.GetMembers(CountPropertyName).OfType<IPropertySymbol>().Single(), true);
```

Comment:
Please switch from `Single` to `FirstOrDefault`. We have had numerous bugs from broken code or references where expected well known types/members are missing or have duplicates.

Reply:
Replaced LINQ with `foreach`.